layers/tcp.go

```diff
@@ -299,9 +307,21 @@ func decodeTCP(data []byte, p gopacket.PacketBuilder) error {
 	if err != nil {
 		return err
 	}
-	return p.NextDecoder(gopacket.LayerTypePayload)
+	if p.DecodeApplicationLayers() {
+		return p.NextDecoder(tcp.NextLayerType())
```

Comment:
Yes, we can rename the function to something more meaningful. Suggestions?

As we see it, for some protocols it does absolutely make sense to use gopacket's existing framework to decode TCP-based protocols in gopacket layers. We primarily deal with industrial protocols which were originally transported over serial lines and are correctly "packetized". This is the purpose of SetConsumed(): to tell the application of much of the payload has already been parsed by the layer, and not resend the same bytes when more data arrives. See also DNS over TCP, which is not stream-based.

For some protocols it is obviously better to use a stream interface, e.g. for HTTP or FTP, and that is what we use. But we see value in both modes, and most of our TCP dissectors use the "packet" mode.